compiler/rustc_data_structures/src/sync.rs

```diff
@@ -360,6 +413,129 @@ cfg_if! {
     }
 }
 
+pub fn for_each_in<T: IntoIterator>(t: T, mut for_each: impl FnMut(T::Item)) {
+    // We catch panics here ensuring that all the loop iterations execute.
+    // This makes behavior consistent with the parallel compiler.
+    let mut panic = None;
+    t.into_iter().for_each(|i| {
+        if let Err(p) = catch_unwind(AssertUnwindSafe(|| for_each(i))) {
+            if panic.is_none() {
+                panic = Some(p);
+            }
+        }
+    });
+    if let Some(panic) = panic {
+        resume_unwind(panic);
+    }
+}
+
+#[macro_export]
+macro_rules! maybe_parallel {
+    ([$tcx: expr], [$tcx_non: expr], $($blocks:expr),*) => {
+        if is_parallel() {
+            parallel!([$tcx], $($blocks),*)
+        } else {
+            // We catch panics here ensuring that all the blocks execute.
+            // This makes behavior consistent with the parallel compiler.
+            let mut panic = None;
+            let tcx = $tcx_non;
+            $(
+                if let Err(p) = ::std::panic::catch_unwind(
+                    ::std::panic::AssertUnwindSafe(|| $blocks(tcx))
+                ) {
+                    if panic.is_none() {
+                        panic = Some(p);
+                    }
+                }
+            )*
+            if let Some(panic) = panic {
+                ::std::panic::resume_unwind(panic);
+            }
+        }
+    }
+}
+
+#[inline]
+pub fn maybe_join<A, B, RA: Send, RB: Send>(oper_a: A, oper_b: B) -> (RA, RB)
+where
+    A: FnOnce() -> RA + DynSendCheck + DynSyncCheck,
+    B: FnOnce() -> RB + DynSendCheck + DynSyncCheck,
+{
+    if is_parallel() {
+        let oper_a = DynSend::new(oper_a);
+        let oper_b = DynSend::new(oper_b);
+        join(move || oper_a.into_inner()(), move || oper_b.into_inner()())
+    } else {
+        (oper_a(), oper_b())
+    }
+}
+
+#[inline]
+pub fn oper<A, B, RA, RB>(oper_a: A, oper_b: B) -> (RA, RB)
+where
+    A: FnOnce() -> RA,
+    B: FnOnce() -> RB,
+{
+    (oper_a(), oper_b())
+}
+
+// Ensure data structures is `Send` by using runtime check.
+// Note: Users need to ensure that the data structure **will not break**
+// thread safety after being created.
+pub unsafe auto trait DynSendCheck {}
+
+// Ensure data structures is `Sync` by using runtime check.
+// Note: Users need to ensure that the data structure **will not break**
+// thread safety after being checked.
+pub unsafe auto trait DynSyncCheck {}
```

Comment:
It would be better to closely mirror `std` naming with `DynSend` and `DynCheck`. These also need the same negative impls as the `std` ones for safety.

Reply:
Done